File: 15_Semantic_Rules/ets_cases/15.14_Compatibility_Features/cross_lang_verify/SEM_15_14_Compatibility.java
/**
 * Java cross-language verification for 15.14 Compatibility Features
 *
 * Key difference: Java requires boolean in conditional contexts;
 * ArkTS allows truthiness (non-boolean values in conditions, &&/|| return operands).
 */
class SEM_15_14_Compatibility {
    static int pass = 0;
    static int fail = 0;

    static void check(boolean cond, String msg) {
        if (cond) { pass++; System.out.println("  PASS " + msg); }
        else { fail++; System.out.println("  FAIL " + msg); }
    }

    public static void main(String[] args) {
        // === Truthiness: Java requires boolean ===
        // ArkTS: if (0) { ... }  // false (0 is falsy)
        // Java:  COMPILE ERROR - int cannot be converted to boolean
        {
            // int x = 0;
            // if (x) { }  // COMPILE ERROR in Java
            check(true, "truthiness: Java requires boolean in if (ArkTS allows non-boolean)");
        }

        // Java equivalent: must write explicit comparison
        {
            int x = 0;
            if (x != 0) {
                check(false, "zero is truthy - should not execute");
            } else {
                check(true, "zero is falsy (Java: explicit comparison)");
            }
        }

        // === Non-empty string truthiness ===
        {
            String s = "hello";
            if (s != null && !s.isEmpty()) {
                check(true, "non-empty string truthiness (Java: explicit null+empty check)");
            }
        }

        // === Empty string falsiness ===
        {
            String s = "";
            if (s != null && !s.isEmpty()) {
                check(false, "empty string should be falsy");
            } else {
                check(true, "empty string falsiness (Java: explicit isEmpty check)");
            }
        }

        // === && and || return boolean in Java ===
        // ArkTS: let r = 0 && "foo"  // r: number|string (returns operand)
        // Java:  boolean r = true && false  // r is always boolean
        {
            boolean r = true && false;
            check(!r, "&& returns boolean in Java (ArkTS returns operand)");
        }

        {
            boolean r = true || false;
            check(r, "|| returns boolean in Java (ArkTS returns operand)");
        }

        // === Truthiness table ===
        {
            // Java: all of these are COMPILE ERRORS in conditionals
            // Must use explicit boolean expressions
            check(!(0 != 0), "zero is falsy (Java: 0 != 0 is false)");
            check(1 != 0, "non-zero is truthy (Java: 1 != 0 is true)");
            check(!("".length() > 0), "empty string falsy (Java: length check)");
            check("x".length() > 0, "non-empty string truthy");
            check(null == null, "null is falsy (Java: null == null is true)");
        }

        // === Ternary expression ===
        // ArkTS: let r = cond ? 42 : "str"  // r: int|string (union type)
        // Java:  Object r = cond ? Integer.valueOf(42) : "str"  // must use common supertype
        {
            Object r = true ? Integer.valueOf(42) : "str";
            check(r instanceof Integer && ((Integer) r).intValue() == 42,
                   "ternary returns common supertype (Java: Object for int|string)");
        }

        System.out.println("\n=== Summary ===");
        System.out.println("Pass: " + pass + "  Fail: " + fail);
        if (fail > 0) { System.exit(1); }
        System.out.println("verified");
    }
}

File: 15_Semantic_Rules/ets_cases/15.14_Compatibility_Features/cross_lang_verify/SEM_15_14_Compatibility.swift
/**
 * Swift cross-language verification for 15.14 Compatibility Features
 *
 * Key difference: Swift requires Bool in conditional contexts;
 * ArkTS allows truthiness (non-boolean values in conditions, &&/|| return operands).
 */
import Foundation

var pass = 0
var fail = 0

func check(_ cond: Bool, _ msg: String) {
    if cond { pass += 1; print("  PASS \(msg)") }
    else { fail += 1; print("  FAIL \(msg)") }
}

// === Truthiness: Swift requires Bool ===
// ArkTS: if (0) { ... }  // false (0 is falsy)
// Swift: COMPILE ERROR - Int cannot be converted to Bool
do {
    // let x = 0
    // if x { }  // COMPILE ERROR in Swift
    check(true, "truthiness: Swift requires Bool in if (ArkTS allows non-boolean)")
}

// Swift equivalent: must write explicit comparison
do {
    let x = 0
    if x != 0 {
        check(false, "zero is truthy - should not execute")
    } else {
        check(true, "zero is falsy (Swift: explicit comparison)")
    }
}

// === Non-empty string truthiness ===
do {
    let s = "hello"
    if !s.isEmpty {
        check(true, "non-empty string truthiness (Swift: explicit isEmpty check)")
    }
}

// === Empty string falsiness ===
do {
    let s = ""
    if s.isEmpty {
        check(true, "empty string falsiness (Swift: explicit isEmpty check)")
    } else {
        check(false, "empty string should be falsy")
    }
}

// === && and || return Bool in Swift ===
// ArkTS: let r = 0 && "foo"  // r: number|string (returns operand)
// Swift:  let r = true && false  // r is always Bool
do {
    let r = true && false
    check(!r, "&& returns Bool in Swift (ArkTS returns operand)")
}

do {
    let r = true || false
    check(r, "|| returns Bool in Swift (ArkTS returns operand)")
}

// === Truthiness table ===
do {
    check(!(0 != 0), "zero is falsy (Swift: 0 != 0 is false)")
    check(1 != 0, "non-zero is truthy (Swift: 1 != 0 is true)")
    check("".isEmpty, "empty string falsy (Swift: isEmpty)")
    check(!"x".isEmpty, "non-empty string truthy")
    check(nil == nil, "nil is falsy (Swift: nil == nil is true)")
}

// === Ternary expression ===
// ArkTS: let r = cond ? 42 : "str"  // r: int|string (union type)
// Swift:  let r = cond ? 42 : 0     // both branches must be same type
do {
    let r = true ? 42 : 0
    check(r == 42, "ternary returns same type (Swift: both branches must match)")
}

print("\n=== Summary ===")
print("Pass: \(pass)  Fail: \(fail)")
if fail > 0 { exit(1) }
print("verified")
